sdk/python/kfp/compiler/pipeline_spec_builder.py

```diff
@@ -1570,6 +1571,135 @@ def write_pipeline_spec_to_file(pipeline_spec: pipeline_spec_pb2.PipelineSpec,
+                task_spec.component_ref.name = component
+
+        # for inner task group calling on components
+        for _, component_spec in pipeline_spec.components.items():
+            if component_spec.dag:
+                for __, task_spec in component_spec.dag.tasks.items():
+                    if task_spec.component_ref.name in clones:
+                        task_spec.component_ref.name = component
+
+    # clean up other component spec names
+    if components_with_clones:
+        changed_names = {}
+        for component, component_spec in pipeline_spec.components.items():
+            if component not in components_with_clones and component_spec.executor_label:
+                last_delimiter = component.rfind('-')
+                if len(component) > last_delimiter + 1 and component[
+                        last_delimiter + 1:].isnumeric():
+                    initial = component
+                    component = component[:-2]
+                    component = utils.make_name_unique_by_adding_index(
+                        name=component,
+                        collection=pipeline_spec.components.keys(),
+                        delimiter='-')
+
+                    executor_spec = pipeline_spec.deployment_spec.fields[
+                        'executors'].struct_value.fields[
+                            component_spec.executor_label]
+                    del pipeline_spec.deployment_spec.fields[
+                        'executors'].struct_value.fields[
+                            component_spec.executor_label]
+
+                    corresponding_executor_name = component_spec.executor_label
+                    corresponding_executor_name = corresponding_executor_name[:
+                                                                              -2]
+                    corresponding_executor_name = utils.make_name_unique_by_adding_index(
+                        name=corresponding_executor_name,
+                        collection=pipeline_spec.deployment_spec
+                        .fields['executors'].struct_value.fields.keys(),
+                        delimiter='-')
+
+                    component_spec.executor_label = corresponding_executor_name
+                    pipeline_spec.deployment_spec.fields[
+                        'executors'].struct_value.fields[
+                            corresponding_executor_name].CopyFrom(executor_spec)
+
+                    changed_names[initial] = [component, component_spec]
+
+                    for _, task_spec in pipeline_spec.root.dag.tasks.items():
+                        if task_spec.component_ref.name == initial:
+                            task_spec.component_ref.name = component
+
+        for initial, new_details in changed_names.items():
+            del pipeline_spec.components[initial]
+            component_name, component_spec = new_details
+            pipeline_spec.components[component_name].CopyFrom(component_spec)
+
```

Comment:
I'm curious if this function can be simplified by maintaining a map of (unique) serialized executor bodies to the executor label we want to use for it. Pseudocode is probably better than a description for conveying the idea:

```python
unique_executors = {}  # maps body to label
for component_name, component_body in components.items():
    if component.executor_label:
        executor_body = executors[executor_label]
        if executor_body in unique_executors:
            executor_body_str = serialize_executor_body(executor_body)
            unique_executors[executor_body] = correct_executor_label
            update_pipeline_spec(pipeline_spec, component_name, executor_label)
```

WDYT?

Reply:
This simplification would change a lot of the logic that has already been written, I don't know if its worth trying to build the logic from scratch on top of this